fsmonitor.c

```diff
@@ -32,17 +46,27 @@ int read_fsmonitor_extension(struct index_state *istate, const void *data,
 	uint32_t ewah_size;
 	struct ewah_bitmap *fsmonitor_dirty;
 	int ret;
+	uint64_t tm;
+	struct strbuf last_update = STRBUF_INIT;
 
 	if (sz < sizeof(uint32_t) + sizeof(uint64_t) + sizeof(uint32_t))
 		return error("corrupt fsmonitor extension (too short)");
 
 	hdr_version = get_be32(index);
 	index += sizeof(uint32_t);
-	if (hdr_version != INDEX_EXTENSION_VERSION)
+	if (hdr_version == INDEX_EXTENSION_VERSION1) {
+		tm = get_be64(index);
+		strbuf_addf(&last_update, "%"PRIu64"", tm);
+		index += sizeof(uint64_t);
+	} else if (hdr_version == INDEX_EXTENSION_VERSION2) {
+		strbuf_addstr(&last_update, index);
+		index += last_update.len + 1;
+	} else {
+		strbuf_release(&last_update);
 		return error("bad fsmonitor version %d", hdr_version);
+	}
 
-	istate->fsmonitor_last_update = get_be64(index);
-	index += sizeof(uint64_t);
+	istate->fsmonitor_last_update = strbuf_detach(&last_update, NULL);
```

Comment:
Just checking:  In the !v1 and !v2 else case, we never put anything in `last_update`.  Here we're stealing/detaching an empty buffer.  I assume that is OK ??

Reply:
The `istate->fsmonitor_last_update` is used as a flag for checking if we are using fsmonitor in some places so we need to keep it something other than `0` for those checks.

Reply:
`strbuf_detach()` is always okay, even if nothing was added to the buffer: [`strbuf_detach()`](https://github.com/git/git/blob/53a06cf39b756eddfe4a2a34da93e3d04eb7b728/strbuf.c#L69-L78) calls [`strbuf_grow()`](https://github.com/git/git/blob/53a06cf39b756eddfe4a2a34da93e3d04eb7b728/strbuf.c#L90-L101) to make sure that _some_ memory is allocated.

Still, if we don't have to detach anything, it's better to avoid allocating something ;-)